operator/controller/src/main/java/io/apicurio/registry/operator/resource/ResourceFactory.java

```diff
@@ -28,16 +38,157 @@ public class ResourceFactory {
     public static final String UI_CONTAINER_NAME = "apicurio-registry-ui";
 
     public Deployment getDefaultAppDeployment(ApicurioRegistry3 primary) {
-        var r = getDefaultResource(primary, Deployment.class, RESOURCE_TYPE_DEPLOYMENT, COMPONENT_APP);
-        addDefaultLabels(r.getSpec().getTemplate().getMetadata().getLabels(), primary, COMPONENT_APP);
+        var r = new Deployment();
+
+        // Metadata
+        if (r.getMetadata() == null) {
+            r.setMetadata(new ObjectMeta());
+        }
+        if (r.getMetadata().getLabels() == null) {
+            r.getMetadata().setLabels(new HashMap<>());
+        }
+        r.getMetadata().setNamespace(primary.getMetadata().getNamespace());
+        r.getMetadata().setName(
+                primary.getMetadata().getName() + "-" + COMPONENT_APP + "-" + RESOURCE_TYPE_DEPLOYMENT);
+        addDefaultLabels(r.getMetadata().getLabels(), primary, COMPONENT_APP);
+
+        // Spec
+        r.setSpec(new DeploymentSpec());
+        r.getSpec().setReplicas(1);
+        r.getSpec().setSelector(new LabelSelector());
         addSelectorLabels(r.getSpec().getSelector().getMatchLabels(), primary, COMPONENT_APP);
+        if (primary.getSpec().getApp().getPodTemplateSpec() != null) {
+            r.getSpec().setTemplate(primary.getSpec().getApp().getPodTemplateSpec());
+        } else {
+            r.getSpec().setTemplate(new PodTemplateSpec());
+        }
+        if (r.getSpec().getTemplate().getMetadata() == null) {
+            r.getSpec().getTemplate().setMetadata(new ObjectMeta());
+        }
+        addDefaultLabels(r.getSpec().getTemplate().getMetadata().getLabels(), primary, COMPONENT_APP);
+        var c = getContainer(r.getSpec().getTemplate(), APP_CONTAINER_NAME);
+        if (c == null) {
+            if (r.getSpec().getTemplate().getSpec() == null) {
+                r.getSpec().getTemplate().setSpec(new PodSpec());
+            }
+            c = new Container();
+            c.setName(APP_CONTAINER_NAME);
+            if (r.getSpec().getTemplate().getSpec().getContainers() == null) {
+                r.getSpec().getTemplate().getSpec().setContainers(new ArrayList<>());
+            }
+            r.getSpec().getTemplate().getSpec().getContainers().add(c);
+        }
+        if (isBlank(c.getImage())) {
+            c.setImage("quay.io/apicurio/apicurio-registry:latest-snapshot");
+        }
+        if (c.getEnv() != null && !c.getEnv().isEmpty()) {
+            throw new OperatorException("""
+                    Field spec.(app/ui).podTemplateSpec.spec.containers[name = %s].env must be empty.  \
+                    Use spec.(app/ui).env to configure environment variables."""
+                    .formatted(APP_CONTAINER_NAME));
+        }
+        if (c.getPorts() == null) {
+            c.setPorts(new ArrayList<>());
+        }
+        mergeNotOverride(c.getPorts(), List.of(new ContainerPortBuilder().withName("http").withProtocol("TCP")
```

Comment:
no need to merge anything here as we are building a brand new thing.

Reply:
the comment applies to all the occurences.

Reply:
I don't follow. We are building on top of PTs from the spec. The only part we're building from scratch is the Deployment that contains the PTS.

Reply:
We're merging the ports for example.

Reply:
I'll follow up in another PR.